build_runner/lib/src/environment/overridable_environment.dart

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) 2018, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:logging/logging.dart';
+
+import '../asset/reader.dart';
+import '../asset/writer.dart';
+import '../generate/directory_watcher_factory.dart';
+import 'build_environment.dart';
+
+/// A [BuildEnvironment] which can have individual features overridden.
+class OverrideableEnvironment implements BuildEnvironment {
+  final BuildEnvironment _default;
+
+  final RunnerAssetReader _reader;
+  final RunnerAssetWriter _writer;
+  final DirectoryWatcherFactory _directoryWatcherFactory;
+
+  final void Function(LogRecord) _onLog;
+
+  OverrideableEnvironment(
+    this._default, {
+    RunnerAssetReader reader,
+    RunnerAssetWriter writer,
+    DirectoryWatcherFactory directoryWatcherFactory,
+    void Function(LogRecord) onLog,
+  })
+      : _reader = reader,
+        _writer = writer,
+        _directoryWatcherFactory = directoryWatcherFactory,
+        _onLog = onLog;
+
+  @override
+  RunnerAssetReader get reader => _reader ?? _default.reader;
+
+  @override
+  RunnerAssetWriter get writer => _writer ?? _default.writer;
+
+  @override
+  DirectoryWatcherFactory get directoryWatcherFactory =>
+      _directoryWatcherFactory ?? _default.directoryWatcherFactory;
+
+  @override
+  void onLog(LogRecord record) {
+    if (_onLog != null) {
```

Comment:
`(onLog ?? _default.onLog)(record)`? I don't know though lol, this is probably fine as is :D

Reply:
Could also be `_onLog?.call(record) ?? _default.onLog(record)` :)

as much as I like minimizing lines like this I think either one masks the intent too much